File: soyeong/programmers-2/오픈채팅방.py
def solution(record):
    answer = []
    uid_list = {}  # dict = uid : nickname

    # 결국 가장 마지막에 기록된 유저 닉네임을 기준으로 출력하는 방식
    # 닉네임 정보 먼저 정리

    for one in record:
        one_list = one.split()
        # 나간거 빼고 닉네임 체크
        if one_list[0] == 'Leave':
            continue
        uid_list[one_list[1]] = one_list[2]

    # 정리된 닉네임을 기준으로 출력
    for chat_one in record:
        chat_list = chat_one.split()
        if chat_list[0] == "Enter":
            answer.append(uid_list[chat_list[1]] + "님이 들어왔습니다.")
        elif chat_list[0] == "Leave":
            answer.append(uid_list[chat_list[1]] + "님이 나갔습니다.")

    return answer

File: kyulee/programmers-2/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        String[] answer = {};
        List<String> result = new ArrayList<>();
        Map<String, String> id = new HashMap<>();

        // id 리스트 작성
        for(String chat : record){
            String[] log = chat.split(" ");
            if(!log[0].equals("Leave")){
                id.put(log[1], log[2]);
            }
        }

        for(String chat : record) {
            String[] log = chat.split(" ");
            String userId = id.get(log[1]);

            if (log[0].equals("Enter")) {
                result.add(userId + "님이 들어왔습니다.");
            } else if (log[0].equals("Leave")) {
                result.add(userId + "님이 나갔습니다.");
            }
        }

        answer = result.toArray(new String[0]);

        return answer;
    }
}